pkg/controller/trainingjob_controller.go

```diff
@@ -0,0 +1,260 @@
+	// recorder is an event recorder for recording Event resources to the
+	// Kubernetes API.
+	recorder record.EventRecorder
+}
+
+// New returns a trainingjob controller
+func New(
+	kubeCli kubernetes.Interface,
+	extCli extclient.Interface,
+	paddleCli paddleclient.Interface,
+	paddleInformers paddleinformers.SharedInformerFactory) *TrainingJobController {
+	trainingjobInformer := paddleInformers.Paddlepaddle().V1().TrainingJobs()
+	paddlescheme.AddToScheme(scheme.Scheme)
+
+	log.Info("Creating trainingjob event broadcaster")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(log.Info)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeCli.CoreV1().Events("")})
+	workqueue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "TrainingJob")
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "TrainingJobController"})
+
+	controller := &TrainingJobController{
+		KubeCli:           kubeCli,
+		ExtCli:            extCli,
+		PaddleCli:         paddleCli,
+		trainingjobLister: trainingjobInformer.Lister(),
+		trainingjobSynced: trainingjobInformer.Informer().HasSynced,
+		jobupdater:        new(sync.Map),
+		workqueue:         workqueue,
+		recorder:          recorder,
+	}
+
+	log.Info("Setting up event handlers")
+	trainingjobInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.enqueue,
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldTj := oldObj.(*paddlev1.TrainingJob)
+			newTj := newObj.(*paddlev1.TrainingJob)
+			objNs := oldTj.Namespace
+			objName := oldTj.Name
+			if oldTj.ResourceVersion == newTj.ResourceVersion {
+				log.Debug("same resourceversion for training job", objNs, "/", objName, ", skipped")
+				return
+			}
+			log.Debug("resourceversion for training job", objNs, "/", objName, "updated")
+			controller.enqueue(newObj)
+		},
+		DeleteFunc: controller.dequeue,
+	})
+
+	return controller
+}
+
+// Run will set up the event handlers for trainingjob, as well as syncing
+// informer caches and starting workers. It will block until stopCh
+// is closed, at which point it will shutdown the workqueue and wait for
+// workers to finish processing their current work items.
+func (c *TrainingJobController) Run(threadiness int, stopCh <-chan struct{}) error {
+	// TODO add a lock to ensure there is only one controller in the cluster
+	defer runtime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	log.Info("Starting to create custom resource definition")
+
+	if err := c.createCRD(); err != nil {
+		return fmt.Errorf("failed to create kind TrainingJob: %v", err)
+	}
+
+	log.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.trainingjobSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	log.Info("Starting workers")
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
```

Comment:
Why do we need a worker pool? Goroutines are already multiplexed onto a thread pool, perhaps we can have a single loop, and start a new goroutine for each iteration.

Reply:
No worries, let me explain.

Here is the current code:

```go
for i := 0; i < threadiness; i++ {
		go wait.Until(c.runWorker, time.Second, stopCh)
}

func (c *TrainingJobController) runWorker() {
	for c.processNextWorkItem() {
	}
}
```

And there are *threadiness* number of gorountines running all work items (assume *N*). The max concurrency is *threadiness*.

Another implementation is:

```go
go wait.Until(c.runWorker, time.Second, stopCh) // only a single wait.Until

func (c *TrainingJobController) runWorker() {
	for item := range c.itemCh {
                go c.process(item)
	}
}
```

In this way the The max concurrency is *N* rather than *threadiness*. Eliminating the unnecessary configuration value *threadiness*, at the same time provides better concurrency.

Reply:
Thanks for your elaborated explanation, and I consulted implementation of some built-in controllers, such as [deployment controller](https://github.com/kubernetes/kubernetes/blob/97fdf8ac70dfdaf07074dee6b6cb345be71e96ec/pkg/controller/deployment/deployment_controller.go#L148) and [job controller](https://github.com/kubernetes/kubernetes/blob/97fdf8ac70dfdaf07074dee6b6cb345be71e96ec/pkg/controller/job/job_controller.go#L137). These controllers usually use a controlled number to indicate the max number of objects that are allowed to sync concurrently, instead of the total number. Besides, sync number can be managed by the [configuration of controller-manager](https://kubernetes.io/docs/reference/generated/kube-controller-manager/), for example `--concurrent-deployment-syncs int32`.
To sum up, I think it's more reasonable to use *threadiness* as the max concurrency.

Reply:
I see, I did not realize that there is a need for limiting the concurrency, now it makes sense. Thanks!